src/metric.cc

```diff
@@ -0,0 +1,324 @@
+      metric_family_address_(metric_address)
+{
+#ifdef TRITON_PB_STUB
+  // Send the request to create the Metric to the parent process
+  std::unique_ptr<Stub>& stub = Stub::GetOrCreateInstance();
+  SaveToSharedMemory(stub->ShmPool());
+  CustomMetricsMessage* custom_metrics_msg = nullptr;
+  try {
+    stub->SendCustomMetricsMessage(
+        &custom_metrics_msg, PYTHONSTUB_MetricRequest, shm_handle_);
+  }
+  catch (const PythonBackendException& pb_exception) {
+    throw PythonBackendException(
+        "Error when creating Metric: " + std::string(pb_exception.what()));
+  }
+  metric_address_ = custom_metrics_msg->address;
+#endif
+}
+
+Metric::~Metric()
+{
+#ifdef TRITON_PB_STUB
+  Clear();
+#endif
+}
+
+const std::string&
+Metric::FamilyName()
+{
+  return family_name_;
+}
+
+const std::string&
+Metric::Labels()
+{
+  return labels_;
+}
+
+bi::managed_external_buffer::handle_t
+Metric::ShmHandle()
+{
+  return shm_handle_;
+}
+
+const MetricRequestKind&
+Metric::RequestKind()
+{
+  return metric_request_kind_;
+}
+
+double
+Metric::Value()
+{
+  return value_;
+}
+
+void
+Metric::SaveToSharedMemory(std::unique_ptr<SharedMemoryManager>& shm_pool)
+{
+  AllocatedSharedMemory<char> custom_metric_shm =
+      shm_pool->Construct<char>(sizeof(MetricShm), true);
+  custom_metric_shm_ptr_ =
+      reinterpret_cast<MetricShm*>(custom_metric_shm.data_.get());
+
+  std::unique_ptr<PbString> family_name_shm =
+      PbString::Create(shm_pool, FamilyName());
+  std::unique_ptr<PbString> labels_shm = PbString::Create(shm_pool, Labels());
+
+  custom_metric_shm_ptr_->value = value_;
+  custom_metric_shm_ptr_->metric_request_kind = metric_request_kind_;
+  custom_metric_shm_ptr_->family_name_shm_handle = family_name_shm->ShmHandle();
+  custom_metric_shm_ptr_->labels_shm_handle = labels_shm->ShmHandle();
+  custom_metric_shm_ptr_->metric_family_address = metric_family_address_;
+  custom_metric_shm_ptr_->metric_address = metric_address_;
+
+  // Save the references to shared memory.
+  custom_metric_shm_ = std::move(custom_metric_shm);
+  family_name_shm_ = std::move(family_name_shm);
+  labels_shm_ = std::move(labels_shm);
+  shm_handle_ = custom_metric_shm.handle_;
+}
+
+std::unique_ptr<Metric>
+Metric::LoadFromSharedMemory(
+    std::unique_ptr<SharedMemoryManager>& shm_pool,
+    bi::managed_external_buffer::handle_t handle)
+{
+  AllocatedSharedMemory<char> custom_metric_shm = shm_pool->Load<char>(handle);
+  MetricShm* custom_metric_shm_ptr =
+      reinterpret_cast<MetricShm*>(custom_metric_shm.data_.get());
+
+  std::unique_ptr<PbString> family_name_shm = PbString::LoadFromSharedMemory(
+      shm_pool, custom_metric_shm_ptr->family_name_shm_handle);
+  std::unique_ptr<PbString> labels_shm = PbString::LoadFromSharedMemory(
+      shm_pool, custom_metric_shm_ptr->labels_shm_handle);
+
+  return std::unique_ptr<Metric>(
+      new Metric(custom_metric_shm, family_name_shm, labels_shm));
+}
+
+Metric::Metric(
```

Comment:
Can you document the difference between the locally loaded metric object (first constructor) vs object created from shared memory handles (second constructor)?

It seems like this class is kind of like 2 classes in 1. First half of the file is local class, and second half is SHM class.

Reply:
Added comments. Let me know if it's still unclear or confusing.